core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/MarecoAllowance.java

```diff
@@ -0,0 +1,294 @@
+                + v1 * rollingStock.getRollingResistanceSecDeriv(v1));
+
+        var u = wle * v1;
+        var du = wleDerivative * v1 + wle;
+        var v = wle + rollingStock.getRollingResistance(v1) * v1;
+        var dv = wleDerivative + rollingStock.getRollingResistanceDeriv(v1) * v1
+                + rollingStock.getRollingResistance(v1);
+        return (du * v - dv * u) / (v * v);
+    }
+
+    @Override
+    protected double getFirstLowEstimate() {
+        return 0;
+    }
+
+    // get the high boundary for the binary search, corresponding to vf = max
+    @Override
+    protected double getFirstHighEstimate(Envelope envelope) {
+        double maxSpeed = envelope.getMaxSpeed();
+
+        double tolerance = .0001; // Stop if you're close enough
+        int maxCount = 200; // Maximum number of Newton's method iterations
+        double x = maxSpeed * 3 / 2; // at high v1 the equation vf = f(v1) tends to be vf = 2*v1/3
+
+        for (int count = 1;
+                //Carry on till we're close, or we've run it 200 times.
+                abs(newtonsMethod(x, maxSpeed)) > tolerance && count < maxCount;
+                count++)  {
+            x = x - newtonsMethod(x, maxSpeed) / newtonsMethodDerivative(x);  //Newtons method.
+        }
+
+        if (abs(newtonsMethod(x, maxSpeed)) <= tolerance) {
+            return x * DICHOTOMY_MARGIN;
+        } else {
+            return maxSpeed * DICHOTOMY_MARGIN; // if no value has been found return a high value to have some margin
+        }
+    }
+
+    @Override
+    protected double getFirstGuess(Envelope envelope) {
+        return (this.getFirstHighEstimate(envelope) + this.getFirstLowEstimate()) / 2;
+    }
+
+    /** Finds the end position of the coasting phases, that will then be generated backwards
+     * with generateCoastingSpeedControllerAtPosition()
+     * Coasting phases can be generated for two different reasons
+     * - a deceleration phase where braking will be replaced by coasting
+     * - before an accelerating slope */
+    private List<Double> findEndOfCoastingPositions(Envelope envelope, double v1) {
+        var res = new ArrayList<Double>();
+        double vf = vf(v1);
+
+        // coasting before deceleration phases
+        var brakingEnvelopeParts = findBrakingEnvelopeParts(envelope);
+        for (var part : brakingEnvelopeParts) {
+            double targetSpeed = part.getEndSpeed();
+            // if that LimitAnnounceSpeedController is above v1 that means it will not have an impact here
+            if (targetSpeed > v1)
+                continue;
+            // deceleration phases that are entirely above vf
+            if (targetSpeed > vf) {
+                res.add(part.getEndPos());
+                continue;
+            }
+            // deceleration phases that cross vf
+            if (part.getMaxSpeed() > vf)
+                res.add(part.interpolatePosition(vf));
```

Comment:
We don't check the return value, it may be null if there's no intersection. Are we sure there's always one?

This would also fail if the envelope isn't bijective. Is it possible to have a small acceleration when braking because of high slopes?

Reply:
@axrolld your attention is required

Reply:
Both assertions are in EnvelopePart method
```java
public Double interpolatePosition(int startIndex, double speed) {
        assert strictlyMonotonicSpeeds;
        assert isBetween(speed, getMinSpeed(), getMaxSpeed());
```

It's not possible to have a small acceleration when braking for safety measures. A train path will never be build on a slope too high for a train to brake because that would represent a serious hazard.